wallet/src/Pos/Util/Mnemonic.hs

```diff
@@ -0,0 +1,664 @@
+        Left "Entropy must be a non-zero sequence of bytes"
+    when (BS.length bytes > 64) $
+        Left "Maximum entropy is 64 bytes (512 bits)"
+    when (BS.length bytes `rem` 4 /= 0) $
+        Left "Entropy must be a multiple of 4 bytes"
+    Right (Entropy bytes)
+
+
+-- | Smart-constructor for the Mnemonic
+mkMnemonic :: [Text] -> Either Text Mnemonic
+mkMnemonic wordsm = do
+    when (length wordsm <= 0) $
+        Left "Mnemonic must be a non-empty list of words"
+    when (length wordsm > 48) $
+        Left "Mnemonic must have a maximum number of 48 words"
+    when (length wordsm `rem` 3 /= 0) $
+        Left "Mnemonic must be a multiple of 3 words"
+    when (isJust $ Text.find (not . isAscii) (unwords wordsm)) $
+        Left "Mnemonic must use only ASCII characters"
+    when (wordsm == map snd (getWords def)) $
+        Left "Forbidden Mnemonic: an example Mnemonic has been submitted. Please generate a fresh and private Mnemonic from a trusted source"
+    indices <- findIndices enWordList wordsm
+    let bits = concatMap word11ToBits indices
+    let (entropyBits, checksum) = splitAt ((length bits `quot` 33) * 32) bits
+    entropy <- mkEntropy (bitsToBytes entropyBits)
+    when (checksum /= calcChecksum entropy) $
+        Left "mnemonic checksum failed: cannot convert mnemonic to entropy"
+    pure $ Mnemonic
+        { getEntropy = entropy
+        , getWords   = zip indices wordsm
+        }
+  where
+    findIndices :: [Text] -> [Text] -> Either Text [Int]
+    findIndices es targets =
+        case mapM (\e -> elemIndex e es) targets of
+        Nothing ->
+            Left "Unknown Mnemonic word(s)"
+        Just is ->
+            pure is
+
+
+-- | An accessor because we don't want to expose the data-type constructor
+entropyToByteString :: Entropy -> ByteString
+entropyToByteString (Entropy e) =
+    e
+
+
+-- | Convert a mnemonic to a seed that can be used to initiate a HD wallet.
+-- Note that our current implementation deviates from BIP-39 as:
+--
+--    - We do not use the password to produce the seed
+--    - We rely on a fast blake2b hashing function rather than a slow PKBDF2
+--
+-- Somehow, we also convert mnemonic to raw bytes using a Blake2b_256 but with
+-- a slightly different approach when converting them to aesKey when redeeming
+-- paper wallets... In this case, we do not serialize the inputs and outputs.
+--
+-- For now, we have two use case for that serialization function. When creating
+-- an HD wallet seed, in which case, the function we use is `serialize'` from
+-- the Pos.Binary module. And, when creating an AESKey seed in which case we
+-- simply pass the `identity` function.
+mnemonicToSeed :: Mnemonic -> ByteString
+mnemonicToSeed =
+    serialize' . blake2b . serialize' . entropyToByteString . mnemonicToEntropy
+
+
+-- | Convert a mnemonic to a seed AesKey. Almost identical to @MnemonictoSeed@
+-- minus the extra serialization.
+mnemonicToAESKey :: Mnemonic -> ByteString
```

Comment:
Well... my two cents is that they _should_ actually be the same function, used with the same purpose :| ... Though, we ended up computing the seed in two different ways in two different part of the code and, dis-tangling the thing yielded these two functions that are nearly identical (the `serialize'` calls away). 

Though, it might be valuable to directly return an `AESKey` here. I'll see whether we can do that without tightening modules too much (Maybe a type-class).